packages/assets-controllers/src/MultichainAssetsController/MultichainAssetsController.ts

```diff
@@ -0,0 +1,562 @@
+    }
+  }
+
+  /**
+   * Updates the assets metadata for the given list of assets
+   *
+   * @param assets - The assets to update
+   */
+  async #updateAssetsMetadata(assets: CaipAssetType[]) {
+    // Creates a mapping of scope to their respective assets list.
+    const assetsByScope: Record<CaipChainId, CaipAssetType[]> = {};
+    for (const asset of assets) {
+      const { chainId } = parseCaipAssetType(asset);
+      if (!assetsByScope[chainId]) {
+        assetsByScope[chainId] = [];
+      }
+      assetsByScope[chainId].push(asset);
+    }
+    let newMetadata: Record<CaipAssetType, FungibleAssetMetadata> = {};
+
+    for (const chainId of Object.keys(assetsByScope) as CaipChainId[]) {
+      const assetsForChain = assetsByScope[chainId];
+      // Now fetch metadata from the associated asset Snaps:
+      const snap = this.#getAssetSnapFor(chainId);
+      if (snap) {
+        const metadata = await this.#getAssetsMetadataFrom(
+          assetsForChain,
+          snap.id,
+        );
+        newMetadata = {
+          ...newMetadata,
+          ...(metadata?.assets ?? {}),
+        };
+      }
+    }
+    this.update((state) => {
+      state.metadata = {
+        ...this.state.metadata,
+        ...newMetadata,
+      };
+    });
+  }
+
+  /**
+   * Creates a mapping of CAIP-2 Chain ID to Asset Snaps.
+   *
+   * @returns A mapping of CAIP-2 Chain ID to Asset Snaps.
+   */
+  #getAssetSnaps(): Record<CaipChainId, Snap[]> {
+    const snaps: Record<CaipChainId, Snap[]> = {};
+    const allSnaps = this.#getAllSnaps();
+    const allPermissions = allSnaps.map((snap) =>
+      this.#getSnapsPermissions(snap.id),
+    );
+
+    for (const [index, permission] of allPermissions.entries()) {
+      let scopes;
+      for (const singlePermissionConstraint of Object.values(permission)) {
+        scopes = getChainIdsCaveat(singlePermissionConstraint);
+        if (!scopes) {
+          continue;
+        }
+        for (const scope of scopes as CaipChainId[]) {
+          if (!snaps[scope]) {
+            snaps[scope] = [];
+          }
+          snaps[scope].push(allSnaps[index]);
+        }
+      }
+    }
+    return snaps;
+  }
+
+  /**
+   * Returns the first asset snap for the given scope
+   *
+   * @param scope - The scope to get the asset snap for
+   * @returns The asset snap for the given scope
+   */
+  #getAssetSnapFor(scope: CaipChainId): Snap | undefined {
+    const allSnaps = this.#snaps[scope];
+    // Pick only the first one, we ignore the other Snaps if there are multiple candidates for now.
+    return allSnaps?.[0]; // Will be undefined if there's no Snaps candidate for this scope.
+  }
+
+  /**
+   * Returns all the asset snaps
+   *
+   * @returns All the asset snaps
+   */
+  #getAllSnaps(): Snap[] {
+    // TODO: Use dedicated SnapController's action once available for this:
+    return this.messagingSystem
+      .call('SnapController:getAll')
+      .filter((snap) => snap.enabled && !snap.blocked);
```

Comment:
```suggestion
    return this.messagingSystem
      .call('SnapController:getRunnableSnaps')
```

Reply:
the snaps-controller exports this type 
```
export type GetAllSnaps = {
    type: `${typeof controllerName}:getAll`;
    handler: SnapController['getAllSnaps'];
};
```
Although, i do not see it exporting the type for the `getRunnableSnaps` ? did i miss something

Reply:
You need to bump `@metamask/snaps-controllers` to `9.19.0`

Reply:
I can bump that on a separate PR 🙏

Reply:
Can we have a ticket to track this ? @sahar-fehri

Reply:
The PR to bump this was already merged; we can make this small update on a separate PR 🙏